tls/extensions/s2n_server_key_share.c

```diff
@@ -35,26 +34,134 @@ const s2n_extension_type s2n_server_key_share_extension = {
     .if_missing = s2n_extension_noop_if_missing,
 };
 
-int s2n_extensions_server_key_share_send_check(struct s2n_connection *conn);
+static int s2n_server_key_share_generate_pq_hybrid(struct s2n_connection *conn, struct s2n_stuffer *out) {
+    notnull_check(out);
+    notnull_check(conn);
+
+    ENSURE_POSIX(s2n_is_in_fips_mode() == false, S2N_ERR_PQ_KEMS_DISALLOWED_IN_FIPS);
 
-static int s2n_server_key_share_send(struct s2n_connection *conn, struct s2n_stuffer *out)
-{
-    GUARD(s2n_extensions_server_key_share_send_check(conn));
+    struct s2n_kem_group_params *server_kem_group_params = &conn->secure.server_kem_group_params;
+
+    notnull_check(server_kem_group_params->kem_group);
+    GUARD(s2n_stuffer_write_uint16(out, server_kem_group_params->kem_group->iana_id));
+
+    struct s2n_stuffer_reservation total_share_size = { 0 };
+    GUARD(s2n_stuffer_reserve_uint16(out, &total_share_size));
+
+    struct s2n_ecc_evp_params *server_ecc_params = &server_kem_group_params->ecc_params;
+    notnull_check(server_ecc_params->negotiated_curve);
+    GUARD(s2n_stuffer_write_uint16(out, server_ecc_params->negotiated_curve->share_size));
+    GUARD(s2n_ecc_evp_generate_ephemeral_key(server_ecc_params));
+    GUARD(s2n_ecc_evp_write_params_point(server_ecc_params, out));
+
+    /* s2n_kem_send_ciphertext() expects to find the client's public key in the
+     * params passed to it. We temporarily point server_kem_params->public_key to
+     * the correct location. */
+    notnull_check(conn->secure.chosen_client_kem_group_params);
+    struct s2n_kem_params *client_kem_params = &conn->secure.chosen_client_kem_group_params->kem_params;
+    struct s2n_kem_params *server_kem_params = &server_kem_group_params->kem_params;
+    notnull_check(client_kem_params->public_key.data);
+    server_kem_params->public_key.data = client_kem_params->public_key.data;
+    server_kem_params->public_key.size = client_kem_params->public_key.size;
+    GUARD(s2n_kem_send_ciphertext(out, server_kem_params));
+    server_kem_params->public_key.data = NULL;
+    server_kem_params->public_key.size = 0;
```

Comment:
Is there a difference from this over just passing in `client_kem_params`?

```
GUARD(s2n_kem_send_ciphertext(out, client_kem_params));
```

Reply:
Short answer - because that's where the shared secret calculation code expects it to be:

https://github.com/awslabs/s2n/blob/1a635cc224d57443f6f584f36b2c3f8ce822fec9/tls/s2n_tls13_handshake.c#L110-L117

Longer answer - this question sent me down an investigative path as to the necessity of `conn->secure.server_kem_group_params` in general. I **think** there isn't any reason for `server_kem_group_params` to necessarily exist. We currently use `server_kem_group_params.kem_group` is multiple places to indicate which KEM group the server chose for negotiation (for example [here](https://github.com/awslabs/s2n/blob/1a635cc224d57443f6f584f36b2c3f8ce822fec9/tls/s2n_tls13_handshake.c#L143-L145) and [here](https://github.com/awslabs/s2n/blob/1a635cc224d57443f6f584f36b2c3f8ce822fec9/tls/extensions/s2n_server_key_share.c#L87-L90)), but I suspect `conn->secure.chosen_client_kem_group_params` can fill that role easily, making `server_kem_group_params` totally redundant. I'd like to keep this as-is for this PR because removing `server_kem_group_params` entirely and transitioning to `chosen_client_kem_group_params` requires some more investigation and is a bit invasive, but I've created https://github.com/awslabs/s2n/issues/2305 to follow up.

Reply:
Talked offline -- it might be clearer why this is happening if we do `GUARD(s2n_kem_send_ciphertext(out, client_kem_params));`, then explicitly copy what we need from client_kem_params to server_kem_params.

Reply:
Will make that change this PR. Longer term, there is definitely room to improve the code quality here and in the shared secret calculation when I get to https://github.com/awslabs/s2n/issues/2305.

Reply:
Actually, there's no reason I can't change where which location we retrieve the PQ secret from when calculating the hybrid shared secret. To avoid an copying, we can pass `client_kem_params` in here, then, in the hybrid shared secret calculation, just retrieve the PQ secret from `chosen_client_kem_group_params->kem_params.shared_secret`.